autogluon/utils/tabular/features/generators/bulk.py

```diff
@@ -0,0 +1,155 @@
+import logging
+from typing import List
+
+import pandas as pd
+from pandas import DataFrame
+
+from .abstract import AbstractFeatureGenerator
+from ..feature_metadata import FeatureMetadata
+
+logger = logging.getLogger(__name__)
+
+
+# TODO: Add parameter to add prefix to each generator to guarantee no name collisions: 'G1_', 'G2_', etc.
+# TODO: Add argument keep_unused, which creates an identity feature generator at each stage to pipe unused input features into the next stage instead of dropping them.
+class BulkFeatureGenerator(AbstractFeatureGenerator):
+    """
+    BulkFeatureGenerator is used for complex feature generation pipelines where multiple generators are required, with some generators requiring the output of other generators as input (multi-stage generation).
+    For ML problems, it is expected that the user uses a feature generator that is an instance of or is inheriting from BulkFeatureGenerator, as single feature generators typically will not satisfy the feature generation needs of all input data types.
+
+    Parameters
+    ----------
+    generators : List[List[AbstractFeatureGenerator]]
+        generators is a list of lists of generators.
+        Feature generators within generators[i] are all fit on the same data, and their outputs are then concatenated to form the output of generators[i].
+        generators[i+1] are then fit on the output of generators[i].
+        The last generator group's output is the output of _fit_transform and _transform methods.
+        Due to the flexibility of generators, at the time of initialization, generators will prepend pre_generators and append post_generators if they are not None.
+            In this case:
+                pre_generators = [[pre_generator] for pre_generator in pre_generators]
+                post_generators = [[post_generator] for post_generator in self._post_generators]
+                self.generators: List[List[AbstractFeatureGenerator]] = pre_generators + generators + post_generators
+                self._post_generators = []
+            This means that self._post_generators will be empty as post_generators will be incorporated into self.generators instead.
+        Note that if generators within a generator group produce a feature with the same name, an AssertionError will be raised as features with the same name cannot be present within a valid DataFrame output.
+            If both features are desired, specify a name_prefix parameter in one of the generators to prevent name collisions.
+            If experimenting with different generator groups, it is encouraged to fit the data frequently during experimentation outside of the ML model training to ensure validity and avoid name collisions.
```

Comment:
unclear what it means to "fit the data frequently during experimentation outside of the ML model training"
Perhaps you mean: "try fitting your experimental feature-generators to the data without any ML model training"

Reply:
Added